src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt

```diff
@@ -0,0 +1,162 @@
+        val uniformMessage = toUniformFieldElement(paddedMessage)
+
+        //second component of the signature
+        val s = uniformMessage.multiply(privateKey).mod(curve.S).add(r).mod(curve.S)
+
+        return Signature(rPoint, s)
+    }
+
+    fun signHashedMessage(msgBytes: ByteArray) : Signature {
+
+        val t = nextBytes(80)
+        val r = hashToScalar(Hash.BLAKE2B, compPersonalization, first = t, second = msgBytes)
+
+        //compute first component of the signature
+        val rPoint = base.scalarMult(r)
+
+        //hash r_x || message
+        val rXCoordBytes = rPoint.x.toByteArray().reversedArray()
+        val paddedMessage = msgBytes + ByteArray(MAX_MESSAGE_SIZE - msgBytes.size)
+
+        val messageDigest = hashToScalar(Hash.BLAKE2S, msgPersonalization, first = rXCoordBytes, second = paddedMessage)
+
+        //second component of the signature
+        val s = messageDigest.multiply(privateKey).mod(curve.S).add(r).mod(curve.S)
+
+        return Signature(rPoint, s)
+    }
+
+    fun verifyRawMessage(msgBytes: ByteArray, signature: Signature) : Boolean {
+        //pad message to max message size and convert to a field element
+        val paddedMessage = msgBytes + ByteArray(MAX_MESSAGE_SIZE - msgBytes.size)
+        val uniformMessage = toUniformFieldElement(paddedMessage)
+
+        val baseMultS = base.scalarMult(signature.s)
+        val pkMultMsg = publicKey.scalarMult(uniformMessage)
+        val rhs = pkMultMsg.add(signature.r)
+
+        return rhs == baseMultS
+    }
+
+    fun verifyHashedMessage(msgBytes: ByteArray, signature: Signature) : Boolean {
+        //hash r_x || message
+        val rXCoordBytes = signature.r.x.toByteArray().reversedArray()
+        val paddedMessage = msgBytes + ByteArray(MAX_MESSAGE_SIZE - msgBytes.size)
+
+        val messageDigest = hashToScalar(Hash.BLAKE2S, msgPersonalization, first = rXCoordBytes, second = paddedMessage)
+
+        //verification
+        val baseMultS = base.scalarMult(signature.s)
+        val pkMultMsg = publicKey.scalarMult(messageDigest)
+        val rhs = pkMultMsg.add(signature.r)
+
+        return rhs == baseMultS
+    }
+
+    private fun hashToScalar(hashFunction: Hash, personalization: ByteArray, first: ByteArray, second:ByteArray) : BigInteger {
+        val digest = when (hashFunction) {
+            Hash.BLAKE2B -> hashBlake2b(first + second, personalization);
+            Hash.BLAKE2S -> hashBlake2s(first.plus(second), personalization)
+        }
+        return toUniformFieldElement(digest)
+    }
+
+    private fun toUniformFieldElement(value: ByteArray): BigInteger = value
+        .reversedArray()
+        .fold(BigInteger.ZERO) { acc, byte ->
+            val ubyte = byte.asUnsigned()
+            (7 downTo 0).fold(acc) { local, shift ->
+                val bit = ubyte shr shift and 1
+                (local + local) % curve.S + if (bit == 1) { BigInteger.ONE } else { BigInteger.ZERO }
```

Comment:
now that I see there is `scalarMult`, it is possible to rewrite as
`local.scalarMult(2) % curve.S + BigInteger.ONE.scalarMult(bit)`

Reply:
`scalarMult()` is an EllipticCurvePoint function. It performs the scalar multiplication on the point, not on the BigIntegers. So, I can't use it in here. But alternatively, I can use one of the following:

`(local + local) % curve.S + BigInteger.valueOf(bit.toLong())`

`(local + local) % curve.S + BigInteger.(bit.toString())`